### ISSUE TITLE
KCP remediation for upgrade and scale up scenarios

### ISSUE THREAD

**Issue by yastij** [kubernetes-sigs/cluster-api#4554]
**User Story**

As a user I would like KCP remediation to work when having 1 CP machine and doing upgrades or scale operations

**Detailed Description**

when upgrading KCP we create a new CP machine, If the bootstrap of the second machine fails the bootstrap script resets the machine through `kubeadm reset`, but we're still left with an instance created. The same goes for scale operations.

Ideally we should enhance the controller's logic to not only rely on the etcd membership but also to look at machines marked for remediation

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature

**Comment by vincepri** [comment]
To clarify, the title mentions upgraden and scale up scenarios (which we should already support), but the description states for a single control plane machine. Is there a reason we'd want to support this? Operations could be potentially disruptive without an operator intervention when the control plane isn't running in HA mode.

**Comment by fabriziopandini** [comment]
Considering the work in https://github.com/kubernetes-sigs/cluster-api/pull/4561, I'm a little bit less concerned about remediation disrupting the cluster, because now there is an improved rule for computing the target quorum and this rule already manages less then etcd members.

The only missing bit is to relax the constraint that KCP should remediate only in N° of replicas > 3, but overall I think this will enable usage of KCP remediation to fix upgrade errors on dev/CI cluster and/or on production use cases where 3 CP are not an option for any reason.

**Comment by yastij** [comment]
agreed, we've done testing and this will be a good a improvement. @fabriziopandini once you open a PR feel free to ping me and I can review
